edb/lib/ext/pgvector.edgeql

```diff
@@ -0,0 +1,189 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2023-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+create extension package pgvector version '1.0' {
+    set ext_module := "ext::pgvector";
+    set sql_extensions := ["vector"];
+
+    create module ext::pgvector;
+
+    create scalar type ext::pgvector::vector extending std::anyscalar {
+        set id := <uuid>"9565dd88-04f5-11ee-a691-0b6ebe179825";
+        set sql_type := "vector";
+        set sql_type_scheme := "vector({__arg_0__})";
+        set num_params := 1;
+    };
+
+    create cast from ext::pgvector::vector to std::str {
+        set volatility := 'Immutable';
+        using sql cast;
+    };
+
+    create cast from std::str to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql cast;
+    };
```

Comment:
My stance is that we should drop this. We don't have `<str>` casts for every format of EdgeDB, we use `to_str()` and `to_datatype()` functions for it. The reason for that is configurability of casting + making sure that we only accept data input compatible with EdgeDB's notion of what's acceptable for the given type.

Given that we have this: `select <pgvector::vector>[1.1, 2.1]` I don't think we need to have string casts at least in 3.0.

Moreover, we don't even know how stable the string casting is in pgvector itself, see https://github.com/pgvector/pgvector/issues/147 as an example.

cc @elvis @msullivan @vpetrovykh

Reply:
Yes, we don't support `str -> array` casts, and neither should we do here.

Reply:
My inclination is to keep the casts, but I don't feel strongly.

Reply:
We don't have a cast from `str` to `array`, so just from that standpoint I would hold off on adding it for `pgvector`.

Reply:
I'm fine with the idea that similar to `array` we skip the direct `str` casts in favour of `to_vector` and `to_str`.

Reply:
At the end of the day the more "natural" cast from a literal is still from an array of numbers of some sort. Which we have. And we also have a cast to `array<float32>`.

Reply:
Alright, I removed the `str` casts. Not adding conversion functions for now because `to_str` wouldn't be able to go into `std`